### ISSUE TITLE
omprog not properly closing child process when signalOnClose=on

### ISSUE THREAD

**Issue by jsiwrk** [rsyslog/rsyslog#2599]
There are three issues with omprog when it is used with the `signalOnClose=on` setting:
* an unresponsive child is not killed within the supposed timeout of 5 secs
* a message "child-process wasn't reaped within timeout" is continously shown in the output
* EOF is not sent to the child

I have found these issues while reviewing the existing omprog tests to try to resolve issue #2403. Below I give more details and my analysis for each of these issues.

When `signalOnClose=on`, omprog sends a TERM signal to the child process and waits for the child to terminate within a ([hardcoded](https://github.com/rsyslog/rsyslog/blob/v8.33.0/plugins/omprog/omprog.c#L64)) timeout of 5 seconds. After this timeout, omprog should supposedly kill the child. However, this does not work.

It is easy to reproduce by executing the existing test `omprog-cleanup-when-unresponsive.sh` with an increased main queue shutdown timeout, e.g. 1 minute: `main_queue(queue.timeoutShutdown="60000")`. The test terminates after 1 minute, not after 5 seconds. By forcing the test to fail, the output in `test-suite.log` shows that omprog finally kills the child, but AFTER the worker thread has been canceled because of the main queue timeout: see [test-suite.log](https://github.com/rsyslog/rsyslog/files/1865601/test-suite.omprog-cleanup-unresponsive.with.increased.queue.timeoutShutdown.log).

My analysis: the auxiliary thread controlling the timeout is not able to [interrupt the waiting thread](https://github.com/rsyslog/rsyslog/blob/v8.33.0/plugins/omprog/omprog.c#L387). The code seems correct, but apparently the signal is being handled by another thread. I have not managed to make it work using an interrupted wait.

Besides, during the shutdown timeout, omprog continously outputs the following message. (Note: I had to truncate the file, since its size was around 200 Mb because of this repeated message.)
```
rsyslog internal message (3,-2428): omprog: child-process wasn't reaped within timeout (5000 ms) preparing to force-kill. [v8.34.0.master try http://www.rsyslog.com/e/2428 ]
rsyslog internal message (3,-2428): omprog: child-process wasn't reaped within timeout (5000 ms) preparing to force-kill. [v8.34.0.master try http://www.rsyslog.com/e/2428 ]
rsyslog internal message (3,-2428): omprog: child-process wasn't reaped within timeout (5000 ms) preparing to force-kill. [v8.34.0.master try http://www.rsyslog.com/e/2428 ]
rsyslog internal message (3,-2428): omprog: child-process wasn't reaped within timeout (5000 ms) preparing to force-kill. [v8.34.0.master try http://www.rsyslog.com/e/2428 ]
...
```

The reason for this repeated message is an absolute timeout not being recalculated [in a pthread_cond_timedwait call](https://github.com/rsyslog/rsyslog/blob/v8.33.0/plugins/omprog/omprog.c#L381).

Regarding the third issue: omprog waits for the child to terminate [without previously closing the pipe](https://github.com/rsyslog/rsyslog/blob/v8.33.0/plugins/omprog/omprog.c#L507-L526), so the child does not receive an EOF. The child should be given an opportunity to terminate on EOF, even if it ignores the TERM signal.

### Environment
- rsyslog version: 8.34.0
- platform: Ubuntu 16.04

**Comment by jsiwrk** [comment]
Regarding this:
>  the auxiliary thread controlling the timeout is not able to interrupt the waiting thread. The code seems correct, but apparently the signal is being handled by another thread.

Unless someone knows how to make this interrupted wait to work, I'll use a busy-sleep wait (repeatedly calling waitpid with WNOHANG). I know this is less optimal, but I haven't managed to make the interrupted wait work, and efficiency doesn't seem crucial when unscheduling a worker (but I could be wrong, please correct me if so...). Besides, using a plain sleep loop greatly simplifies the code, and allows it to be compiled in all platforms I think (the timeout feature is currently being compiled for linux only). It seems a good thing that omprog can kill unresponsive child processes in all platforms.

**Comment by jsiwrk** [comment]
Assigning to myself since I'm including a fix for this issue as part of PR #2629.
